core/src/main/scala/org/http4s/headers/package.scala

```diff
@@ -5,16 +5,35 @@ import org.http4s.util.Writer
 
 package object headers {
 
-  private[headers] def xB3RenderValueImpl(writer: Writer, id: Long): writer.type =
-    if (id == 0L) {
-      val idLength = 16
-      writer.append("0" * idLength)
-    } else {
-      val bitsToHalfByteRatio = 4
-      val leadingHexZeroCount =
-        JLong.numberOfLeadingZeros(id) / bitsToHalfByteRatio
-      writer
-        .append("0" * leadingHexZeroCount)
-        .append(id.toHexString)
+  private[headers] def xB3RenderValueImpl(writer: Writer, idMostSigBits: Long, idLeastSigBits: Option[Long] = None): writer.type = {
+    val bitsToHalfByteRatio = 4
+    val idLength = 16
+    idLeastSigBits match {
+      case Some(idLsb) =>
+        if (idMostSigBits == 0L && idLsb == 0L) {
+          writer.append("0" * idLength * 2)
```

Comment:
We could do the tiniest of microoptimizations here, and below with `"0" * idLength`, by caching these strings.  Might even be faster to `.take` from the longer of these on the other zeroes.

This is not a blocker, just a thought.